### PR TITLE
Recenter 

### LINKED ISSUE #93: Recenter images if needed for wavefront maintenance
### Description
Need functionality in QUIP to center the image at the desired WAS location when not doing LOS Jitter at the begging of the Routine Fine Phasing observations.  The "autoQuip" mode would also need to create the QUIP Out file for WEx to process then send those updated images to WAS Automatic mode instead of ones from DMS.  So this is a feature request for “auto-quip” (which is just a pre-processing routine that didn’t use quip (ginga) at all) to center the target at the WAS-expected point before invoking the WAS.

### PR DESCRIPTION
**Pull request by kulpster85**
### Description
Updates to QUIP to support routine wavefront maintenance. Some of the routine observations have shown that we may need to recenter the image since the WAS is particular about the centroid of the image to fall on a XY of NRCA3 FP3 (464, 1412). So now if QUIP is run with WAVEFRONT_MAINTENANCE operation then it will invoke the recenter function to do some batch processing and update the quip_out.xml if the centroid was offset by 10px in X or Y.

This pull request is to address ...

Fixes #93

### CODE REVIEW

File: wss_tools/utils/recenter.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
+''' Module to recenter images '''
+import os
+from astropy.io import fits
+import matplotlib.pyplot as plt
```
Comment (pllim) [inline wss_tools/utils/recenter.py:4]: matplotlib is not a required dependency. So, we might have to add it here:

https://github.com/spacetelescope/wss_tools/blob/b99e9341afc3bffcbafa6af8f450ea69ded6e569/setup.cfg#L38

File: wss_tools/utils/recenter.py [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+import numpy as np
+from scipy import ndimage
+
```
Comment (pllim) [inline wss_tools/utils/recenter.py:7]:
```suggestion

__all__ = ['rebin', 'recenter']
```

**Review by pllim — COMMENTED** [review]
Aside from implementation discussions offline, I have some more comments here.

API doc needs to be added to https://github.com/spacetelescope/wss_tools/blob/master/docs/wss_tools/ref_api.rst .

Would be nice to have unit tests to go with recentering but if the scientist did not write any for you, I guess there is little you can do here unless you are compelled to write them yourself.

Need a change log in `CHANGES.rst`.

Too many magic numbers. Can we make them less magical?

File: wss_tools/utils/recenter.py (hunk 1 above)
Comment (kulpster85) [reply wss_tools/utils/recenter.py:4]: OK I will add

File: wss_tools/utils/recenter.py (hunk 2 above)
Comment (kulpster85) [reply wss_tools/utils/recenter.py:7]: I will replace the import lines with suggested change

File: wss_tools/utils/recenter.py (hunk 1 above)
Comment (kulpster85) [reply wss_tools/utils/recenter.py:4]: I will update the ref_api.rest using sphinx.  I considered adding a recenter test - I can consult with @nflagey-stsci on how best to do this.

File: wss_tools/utils/recenter.py [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+        # Rebin image with very big pixels to find the approximate location
+        size = 32
+        rebindata = rebin(data, [size, size])
+
+        # Remove background
+        rebindata -= np.median(rebindata)
+
+        # Find the center of mass and cut a box around: we should see the whole
+        # PSF with about 1-2 PSF wide margins
+        com = ndimage.center_of_mass(rebindata)
```
Comment (nflagey-stsci) [inline wss_tools/utils/recenter.py:46]: margin_right = 2.5
margin_left = 1.5
margin_top = 2.5
margin_bottom = 1.5

File: wss_tools/utils/recenter.py [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+        subdata = data[int((com[0] - 1.5) * 2048 / size):
+                       int((com[0] + 2.5) * 2048 / size),
+                       int((com[1] - 1.5) * 2048 / size):
+                       int((com[1] + 2.5) * 2048 / size)]
```
Comment (nflagey-stsci) [inline wss_tools/utils/recenter.py:50]:
```suggestion
        margin_left = 1.5
        margin_right = 2.5
        margin_bottom = 1.5
        margin_top = 2.5
        subdata = data[int((com[0] - margin_left) * imsize / size):
                       int((com[0] + margin_right) * imsize / size),
                       int((com[1] - margin_bottom) * imsize / size):
                       int((com[1] + margin_top) * imsize / size)]
```

File: wss_tools/utils/recenter.py [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+        # Remove background
+        rebinsubdata -= np.median(rebinsubdata)
+
+        # Find the center of mass and cut a box around: should see the inside
+        # of the PSF
+        subdatacom = ndimage.center_of_mass(rebinsubdata)
+        subdata2 = subdata[int((subdatacom[0] - .5) * 2048 / size * 4 / size):
+                           int((subdatacom[0] + 1.5) * 2048 / size * 4 / size),
+                           int((subdatacom[1] - .5) * 2048 / size * 4 / size):
+                           int((subdatacom[1] + 1.5) * 2048 / size * 4 / size)]
```
Comment (nflagey-stsci) [inline wss_tools/utils/recenter.py:64]:
```suggestion
        margin_left2 = 0.5
        margin_right2 = 1.5
        margin_bottom2 = 0.5
        margin_top2 = 1.5
        subdata2 = subdata[int((subdatacom[0] -margin_left2) * imsize / size * 4 / size):
                           int((subdatacom[0] + margin_right2) * imsize / size * 4 / size),
                           int((subdatacom[1] -margin_bottom2) * imsize / size * 4 / size):
                           int((subdatacom[1] + margin_top2) * imsize / size * 4 / size)]
```

File: wss_tools/utils/recenter.py [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        # Find the center of mass
+        com2 = ndimage.center_of_mass(subdata2)
+
+        # Recenter the image to 464, 1412 instead to match the WAS expectations
+        xcpsf = com2[0] + int((subdatacom[0] - .5) * 2048 / size * 4 / size)\
+                        + int((com[0] - 1.5) * 2048 / size)
+        ycpsf = com2[1] + int((subdatacom[1] - .5) * 2048 / size * 4 / size) \
+                        + int((com[1] - 1.5) * 2048 / size)
+        offsetdata = np.roll(data, (464 - int(ycpsf), 1412 - int(xcpsf)),
+                             axis=(1, 0))
```
Comment (nflagey-stsci) [inline wss_tools/utils/recenter.py:75]:
```suggestion
        xcenter = 464
        ycenter = 1412
        offsetdata = np.roll(data, (xcenter - int(ycpsf), ycenter - int(xcpsf)),
                             axis=(1, 0))
```

File: wss_tools/utils/recenter.py [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+
+        # Save back image into file
+        if abs(464 - ycpsf) > 10 or abs(1412 - xcpsf) > 10:
+            # Do the same for the ERR and the DQ
+            hdul[2].data = np.roll(hdul[2].data,
+                                   (464 - int(ycpsf), 1412 - int(xcpsf)),
+                                   axis=(1, 0))
+            hdul[3].data = np.roll(hdul[3].data,
+                                   (464 - int(ycpsf), 1412 - int(xcpsf)),
+                                   axis=(1, 0))
```
Comment (nflagey-stsci) [inline wss_tools/utils/recenter.py:85]:
```suggestion
        if abs(xcenter - ycpsf) > 10 or abs(ycenter - xcpsf) > 10:
            # Do the same for the ERR and the DQ
            hdul[2].data = np.roll(hdul[2].data,
                                   (xcenter - int(ycpsf), ycenter - int(xcpsf)),
                                   axis=(1, 0))
            hdul[3].data = np.roll(hdul[3].data,
                                   (xcenter - int(ycpsf), ycenter - int(xcpsf)),
                                   axis=(1, 0))
```

File: wss_tools/utils/recenter.py [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+                           int((subdatacom[1] + 1.5) * 2048 / size * 4 / size)]
+
+        # Find the center of mass
+        com2 = ndimage.center_of_mass(subdata2)
+
+        # Recenter the image to 464, 1412 instead to match the WAS expectations
+        xcpsf = com2[0] + int((subdatacom[0] - .5) * 2048 / size * 4 / size)\
+                        + int((com[0] - 1.5) * 2048 / size)
+        ycpsf = com2[1] + int((subdatacom[1] - .5) * 2048 / size * 4 / size) \
+                        + int((com[1] - 1.5) * 2048 / size)
```
Comment (nflagey-stsci) [inline wss_tools/utils/recenter.py:73]:
```suggestion
        xcpsf = com2[0] + int((subdatacom[0] - margin_left2) * imsize / size * 4 / size)\
                        + int((com[0] - margin_left) * imsize / size)
        ycpsf = com2[1] + int((subdatacom[1] - margin_bottom2) * imsize / size * 4 / size) \
                        + int((com[1] - margin_bottom) * imsize / size)
```

File: wss_tools/utils/recenter.py [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+    :param outputdir:
+    :param doplot:
+    :return: List of output_images that have been recentered
+    '''
+    output_images = []
+    for im_fn in images:
+        # Open fits
+        hdul = fits.open(im_fn)
+        data = hdul[1].data
+
```
Comment (nflagey-stsci) [inline wss_tools/utils/recenter.py:36]: We could test that the image is actually 2048 by 2048, but let's assume that will always be the case:
```suggestion
imsize = 2048
```

File: wss_tools/utils/recenter.py (hunk 2 above)
Comment (kulpster85) [reply wss_tools/utils/recenter.py:7]: I still see imports in other python modules within utils: mosaic and io - how does flake8 handle this?  In my local flake8 it flagged errors when I switched to __all__

Comment (pllim) [reply wss_tools/utils/recenter.py:7]: Not sure what you mean, the flake8 CI is green. Is this still an issue for you?

Comment (pllim) [reply wss_tools/utils/recenter.py:7]: You need the `__all__` defined for Sphinx to pick up API doc properly.

Comment (kulpster85) [reply wss_tools/utils/recenter.py:7]: OK I added __all__ I think I misunderstood the suggested change initially.

File: wss_tools/utils/recenter.py [hunk 10]
```diff
@@ -0,0 +1,142 @@
...
+from astropy.io import fits
+import matplotlib.pyplot as plt
+import numpy as np
+from scipy import ndimage
+
+__all__ = ['rebin', 'recenter']
+
+
+def rebin(arr, new_shape):
+    ''' rebin(arr, new_shape)
```
Comment (pllim) [inline wss_tools/utils/recenter.py:12]: There is no need to manually type out the func sig. Sphinx does that for you. Instead, you should concisely describe the purpose of this function.

File: wss_tools/utils/recenter.py [hunk 11]
```diff
@@ -0,0 +1,142 @@
...
+    Parameters
+    ----------
+    arr: 2D array
+    new_shape: dimensions of new shape
+    Returns
+    ----------
+    newarr - Rebin 2D array arr to shape new_shape by averaging
```
Comment (pllim) [inline wss_tools/utils/recenter.py:19]:
```suggestion
    arr : ndarray
        2D array.
    new_shape : tuple of int
        Dimensions of new shape.

    Returns
    --------
    newarr : ndarray
        Rebinned array.
```

File: wss_tools/utils/recenter.py (hunk 10 above)
Comment (pllim) [inline wss_tools/utils/recenter.py:12]:
```suggestion
    '''Rebin 2D array to given shape by averaging.
```

Comment (pllim) [reply wss_tools/utils/recenter.py:12]: Also maybe a question for @nflagey-stsci , why not just use https://docs.astropy.org/en/stable/api/astropy.nddata.block_reduce.html here? There is no need to rewrite rebinning.

**Review by pllim — COMMENTED** [review]
Thanks for the changes.

Sphinx docstring could use polishing. Hopefully you can use my suggestions for one function and model it for the other one.

File: wss_tools/utils/recenter.py (hunk 10 above)
Comment (kulpster85) [reply wss_tools/utils/recenter.py]: Thanks for the suggestions

**Comment by kulpster85** [comment]
@pllim @obi-wan76 said that he defers his reveiw to @nflagey-stsci

File: wss_tools/utils/recenter.py [hunk 12]
```diff
@@ -0,0 +1,151 @@
...
+import os
+from astropy.io import fits
+import matplotlib.pyplot as plt
+import numpy as np
+from scipy import ndimage
+
+__all__ = ['rebin', 'recenter']
+
+
+def rebin(arr, new_shape):
```
[the remark this answers is not in the document]
Comment (nflagey-stsci) [reply wss_tools/utils/recenter.py:11]: the usual rebin is using a MEAN while I am using a MEDIAN ... but maybe I did not look enough for a median rebin ... if you know of one, I agree we should switch

[the remark this answers is not in the document]
Comment (pllim) [reply wss_tools/utils/recenter.py:11]: What about https://docs.astropy.org/en/stable/api/astropy.nddata.block_reduce.html#astropy.nddata.block_reduce ? You can pass in your own function.

File: wss_tools/utils/recenter.py [hunk 13]
```diff
@@ -0,0 +1,131 @@
...
+from astropy.nddata import block_reduce
+import matplotlib.pyplot as plt
+import numpy as np
+from scipy import ndimage
+
+__all__ = ['recenter']
+
+
+def recenter(images, outputdir, doplot=False):
+    ''' Recenter images based on NIRCam XY (464,1412) if offset > 10px
```
Comment (pllim) [inline wss_tools/utils/recenter.py:13]:
```suggestion
    '''Recenter images based on NIRCam XY (464,1412) if offset > 10px
```

File: wss_tools/utils/recenter.py [hunk 14]
```diff
@@ -0,0 +1,131 @@
...
+    Parameters
+    ----------
+    images: List
+        List of input images to analyze
+    outputdir: String
+        Working directory where QUIP will write the files to
+    doplot: boolean
+        Show plots to the user via a popup
```
Comment (pllim) [inline wss_tools/utils/recenter.py:21]:
```suggestion
    images : list
        List of input images to analyze.
    outputdir : str
        Working directory where QUIP will write the files to.
    doplot : bool
        Show plots to the user via a popup.
```

File: wss_tools/utils/recenter.py [hunk 15]
```diff
@@ -0,0 +1,131 @@
...
+
+    Returns
+    ----------
+    output_images: List
+        Output images that have been read and/or modified
```
Comment (pllim) [inline wss_tools/utils/recenter.py:26]:
```suggestion
    Returns
    -------
    output_images : list
        Output images that have been read and/or modified.
```

File: wss_tools/utils/recenter.py [hunk 16]
```diff
@@ -0,0 +1,131 @@
...
+    '''
+    output_images = []
+    for im_fn in images:
+        # Open fits
+        hdul = fits.open(im_fn)
```
Comment (pllim) [inline wss_tools/utils/recenter.py:31]: I would strongly recommend doing all the operations related to the FITS data inside a context manager to ensure the file pointer is freed properly on exit.

```python
with fits.open(im_fn) as hdul:
    # code here
```

Or at the very least, add a `hdul.close()` when you don't need it anymore. (No need to call `hdul.close()` if you go the context manager route.)

File: wss_tools/utils/recenter.py [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+        com1 = ndimage.center_of_mass(rebinsubdata)
+        margin_left2 = 0.5
+        margin_right2 = 1.5
+        margin_bottom2 = 0.5
+        margin_top2 = 1.5
+        subdata2 = \
+            subdata[int((com1[0] - margin_left2) * imsize / size * 4 / size):
+                    int((com1[0] + margin_right2) * imsize / size * 4 / size),
+                    int((com1[1] - margin_bottom2) * imsize / size * 4 / size):
+                    int((com1[1] + margin_top2) * imsize / size * 4 / size)]
```
Comment (pllim) [inline wss_tools/utils/recenter.py:72]: `\` continuation has been discouraged for a long time.
```suggestion
        subdata2 = subdata[int((com1[0] - margin_left2) * imsize / size * 4 / size):
                           int((com1[0] + margin_right2) * imsize / size * 4 / size),
                           int((com1[1] - margin_bottom2) * imsize / size * 4 / size):
                           int((com1[1] + margin_top2) * imsize / size * 4 / size)]
```

File: wss_tools/utils/recenter.py [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+        xcntr = 464
+        ycntr = 1412
+
+        # Recenter the image to 464, 1412 instead to match the WAS expectations
+        xcpsf = com2[0] + int((com1[0] - margin_left2)
+                              * imsize / size * 4 / size)\
+                        + int((com[0] - margin_left) * imsize / size)
+        ycpsf = com2[1] + int((com1[1] - margin_bottom2)
+                              * imsize / size * 4 / size) \
+                        + int((com[1] - margin_bottom) * imsize / size)
```
Comment (pllim) [inline wss_tools/utils/recenter.py:86]: Hard to read with weird indentations.
```suggestion
        xcpsf = (com2[0] + int((com1[0] - margin_left2) * imsize / size * 4 / size)
                 + int((com[0] - margin_left) * imsize / size))
        ycpsf = (com2[1] + int((com1[1] - margin_bottom2) * imsize / size * 4 / size)
                 + int((com[1] - margin_bottom) * imsize / size))
```

**Review by pllim — COMMENTED** [review]
I don't grok all the math. I'll just assume TEL has tested this and knows it works. Some technical review as follows.

`doplot` is a little misleading. The plot is "done" no matter what. The flag only controls whether the plot is popped out on the screen or saved to a file. In the past, when I had to implement such a feature, I let user provide the image filename instead and pops out the plot if no filename is given, though it really depends on how you want to use this function (you might want to pop it out *and* save it at the same time). At the very least, for your future sanity, considering renaming to `show_plot`.

Still too many magic numbers but since you are the only user, I'll let this one go...

**Review by pllim — COMMENTED** [review]
There are also formatting problem with your docstring.

```
wss_tools/utils/recenter.py:docstring of wss_tools.utils.recenter.recenter:6: ERROR: Unexpected indentation.
wss_tools/utils/recenter.py:docstring of wss_tools.utils.recenter.recenter:7: WARNING: Block quote ends without a blank line; unexpected unindent.
```

Maybe try to build the doc locally and make sure it succeeds without warnings first?

```
pip install -e .[docs]
cd docs
make html
```

**Pull request merged**